modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Enrich.scala

```diff
@@ -219,4 +232,51 @@ object Enrich {
         .getBytes(UTF_8)
     } else originalBytes
   }
+
+  /**
+   * This is the machinery needed to make sure that no chunk is sunk without
+   * being checkpointed when the app terminates.
+   *
+   * The stream runs on a separate fiber so that we can manually handle SIGINT.
+   *
+   * We use a queue as a level of indirection between the stream of enriched events and the sink + checkpointing.
+   * When we receive a SIGINT or exception then we terminate the fiber by pushing a `None` to the queue.
+   *
+   * The stream is only cancelled after the sink + checkpointing have been allowed to finish cleanly.
+   * We must not terminate the source any earlier, because this would shutdown the kinesis scheduler too early,
+   * and then we would not be able to checkpoint the outstanding records.
+   */
+  private def runWithShutdown[F[_]: Concurrent: Sync: Timer, A](
+    enriched: Stream[F, List[(A, Result)]],
+    sinkAndCheckpoint: Pipe[F, List[(A, Result)], Unit]
+  ): F[Unit] =
+    Queue.synchronousNoneTerminated[F, List[(A, Result)]].flatMap { queue =>
+      queue.dequeue
+        .through(sinkAndCheckpoint)
+        .concurrently(enriched.evalMap(x => queue.enqueue1(Some(x))).onFinalize(queue.enqueue1(None)))
+        .compile
+        .drain
+        .start
+        .bracketCase(_.join) {
+          case (_, ExitCase.Completed) =>
+            // The source has completed "naturally", e.g. processed all input files in the directory
+            Sync[F].unit
+          case (fiber, ExitCase.Canceled) =>
+            // SIGINT received. We wait for the enriched events already in the queue to get sunk and checkpointed
+            terminateStream(queue, fiber)
+          case (fiber, ExitCase.Error(e)) =>
+            // Runtime exception in the stream of enriched events.
+            // We wait for the enriched events already in the queue to get sunk and checkpointed.
+            // We then raise the original exception
+            terminateStream(queue, fiber).attempt *> Logger[F].error(e)("Unexpected error in enrich") *> Sync[F].raiseError(e)
+        }
+    }
+
+  private def terminateStream[F[_]: Concurrent: Sync: Timer, A](queue: NoneTerminatedQueue[F, A], fiber: Fiber[F, Unit]): F[Unit] =
+    for {
+      timeout <- Sync[F].pure(5.minutes)
```

Comment:
Why was this 5 minute timeout needed?  I was expecting termination to work smoothly without any explicit timeout like this.

Reply:
Oh - ignore this comment.  I mis-read it as sleeping for 5 minutes.  This is fine.